### PR TITLE
More granularity analysis in PDO and PDOStatement and granularity type analysis for ParameterType/ArrayParameterType of value

### PR DESCRIPTION
**Pull request by kang8**
|      Q       |   A
|------------- | -----------
| Type         | improvement

#### Summary

Same as #5835, but rebase into 3.6.x.

### CODE REVIEW

File: src/Connection.php [hunk 1]
```diff
@@ -1883,8 +1886,10 @@ private function expandArrayParameters(string $sql, array $params, array $types)
     }
 
     /**
-     * @param array<int, mixed>|array<string, mixed>                               $params
-     * @param array<int, int|string|Type|null>|array<string, int|string|Type|null> $types
+     * @param array<int, mixed>|array<string, mixed>                                         $params
+     * @param array<int, ArrayParameterTypeOfValue>|array<string, ArrayParameterTypeOfValue> $types
+     *
+     * @psalm-assert-if-false array<int, ParameterTypeOfValue>|array<string, ParameterTypeOfValue> $types
```
Comment (kang8) [inline src/Connection.php:1892]: Wrong analysis of psalm.

Except: `array<int|string, 0|1|2|3|5|16|17|Doctrine\DBAL\Types\Type|null|string>`

Actual: `array<int|string, 0|1|2|3|5|16|17|Doctrine\DBAL\Types\Type|ParameterTypeOfValue|null|string>`

I think this is psalm's bug, but I didn't reproduce it, see https://psalm.dev/r/a6a1a210ef.

Tracking this bug.

PS: PHPStan did not report this bug.

File: src/ArrayParameterType.php [hunk 2]
```diff
@@ -2,6 +2,10 @@
 
 namespace Doctrine\DBAL;
 
+/**
+ * @psalm-import-type ParameterTypeOfValue from ParameterType
+ * @psalm-type ArrayParameterTypeOfValue self::*|ParameterTypeOfValue
```
Comment (kang8) [inline src/ArrayParameterType.php:7]: ArrayParameterTypeOfValue is a bad naming. It should be an array parameter type, but in fact it's a mix of array parameter type or parameter type.

My suggestion:

ParameterTypeOfValue -> SingleParameterTypeOfValue
ArrayParameterTypeOfValue -> SingleOrArrayParameterTypeOfValue

**Review by kang8 — COMMENTED** [review]
Some thoughts and items to be completed

File: src/ArrayParameterType.php (hunk 2 above)
Comment (derrabus) [reply src/ArrayParameterType.php:7]: It is and we already have types for this on 4.0.x. I'm working on backporting those.

**Comment by derrabus** [comment]
A small subset of your changes has been merged as #5883 to 3.5 and 3.6. Your original issue has been solved, I suppose. Do you still want to continue with this PR? I'm honestly not sure it's worth it, given we already have better types documented on the 4.0.x branch.

**Comment by kang8** [comment]
OK, thank you to resolve it. Close it.

**Pull request closed without merging**